### PR TITLE
Changed BpfFilterWrapper to use smart pointers internally.

### PR DESCRIPTION
**Pull request by Dimi1010**
Added custom deleter type `BpfProgramDeleter` to encapsulate cleanup logic for a `bpf_program` pointer.

Changed `BpfFilterWrapper` to keep a `unique_ptr` with custom deleter instead of raw pointer to offload the cleanup logic to the smart pointer instead of handling it manually.

### CODE REVIEW

File: Pcap++/src/PcapFilter.cpp [hunk 1]
```diff
@@ -63,8 +67,9 @@ bool BpfFilterWrapper::setFilter(const std::string& filter, LinkLayerType linkTy
 			return false;
 		}
 
+		// TODO: Do we technically need 'free_program' here? Both m_program and m_filter string are overwritten directly after that.
```
Comment (tigercosmos) [inline Pcap++/src/PcapFilter.cpp:70]: I agree that we should remove this.

File: Pcap++/header/PcapFilter.h [hunk 2]
```diff
@@ -78,7 +90,7 @@ namespace pcpp
 	private:
 		std::string m_FilterStr;
 		LinkLayerType m_LinkType;
-		bpf_program* m_Program;
+		std::unique_ptr<bpf_program, detail::BpfProgramDeleter> m_Program;
```
Comment (tigercosmos) [inline Pcap++/header/PcapFilter.h:93]: love this.

**Review by tigercosmos — APPROVED** [review]
Thanks, LGTM.

File: Pcap++/header/PcapFilter.h [hunk 3]
```diff
@@ -69,6 +69,18 @@ namespace pcpp
 		LESS_OR_EQUAL
 	} FilterOperator;
 
+	namespace detail
```
Comment (seladb) [inline Pcap++/header/PcapFilter.h:72]: maybe rename this to `internal` as we have in [other places](https://github.com/seladb/PcapPlusPlus/blob/4d48615c9bfb2ef0c61d50e1e1dd3e3d63a82be0/Common%2B%2B/header/IpUtils.h#L59)?

File: Pcap++/src/PcapFilter.cpp (hunk 1 above)
Comment (seladb) [reply Pcap++/src/PcapFilter.cpp:70]: Yes I agree also

File: Pcap++/header/PcapFilter.h (hunk 3 above)
Comment (tigercosmos) [reply Pcap++/header/PcapFilter.h:72]: We also have `detail` here: https://github.com/seladb/PcapPlusPlus/blob/4d48615c9bfb2ef0c61d50e1e1dd3e3d63a82be0/Pcap%2B%2B/header/PcapFilter.h#L542

it's better to unify them.

Comment (seladb) [reply Pcap++/header/PcapFilter.h:72]: yes I agree! I personally like `internal` better than `detail`

Comment (Dimi1010) [reply Pcap++/header/PcapFilter.h]: Implemented 539201b

File: Pcap++/src/PcapFilter.cpp (hunk 1 above)
Comment (Dimi1010) [reply Pcap++/src/PcapFilter.cpp]: Removed the call in df9f37e. Would still like an opinion on the previous comment tho.

File: Pcap++/src/PcapFilter.cpp [hunk 4]
```diff
@@ -28,15 +28,32 @@ bool GeneralFilter::matchPacketWithFilter(RawPacket* rawPacket)
...
+	}
+
+	/**
+	 * @class PcapTDeleter
+	 * A deleter that cleans up a pcap_t structure by calling pcap_close.
+	 */
+	struct PcapTDeleter
+	{
+		void operator()(pcap_t* ptr) const { pcap_close(ptr); }
+	};
```
Comment (seladb) [inline Pcap++/src/PcapFilter.cpp:46]: This is nice! I think we can use it in more places in the code, especially in `PcapLiveDevice` and `PcapFileDevice.h`. But for now we can leave this code here

**Review by seladb — APPROVED** [review]

**Comment by seladb** [comment]
@tigercosmos can you also re-review before we merge?

File: Pcap++/src/PcapFilter.cpp (hunk 1 above)
Comment (tigercosmos) [reply Pcap++/src/PcapFilter.cpp]: current way is fine.

**Comment by tigercosmos** [comment]
thanks @Dimi1010

**Pull request merged**